client/www/_posts/pg_upgrade.md

```diff
@@ -0,0 +1,474 @@
+    ; ...
+    (wal/start-worker {:conn-config
+                      (or (config/get-next-aurora-config)
+                          ;; Use the next db so that we don't
+                          ;; have to worry about restarting the
+                          ;; invalidator when failing over to a
+                          ;; new db.
+                          (config/get-aurora-config))})
+    ; ...
+    ))
+```
+
+Once we deployed this change, sync servers replicated from 16. Phew, this was at least one step in the story that didn’t feel nerve-wracking!
+
+# 3) Switching Writes
+
+Now to worry about writes:
+
+![](/posts/pg_upgrade/switch_writes.png)
+
+Ultimately, we needed to click some button and trigger a switch. To make the switch work, we’d need to follow two rules:
+
+1. **16 must be caught up**
+    
+    If there are _any_ writes in 13 that haven’t replicated to 16 yet, we can’t turn on writes to 16. Otherwise transactions would come in the wrong order
+
+1. **Once caught up, all new writes must go to 16**
+    
+    If _any_ write accidentally goes to 13, we could lose data.
+
+So, how could we follow these rules?
+
+## We could stop the world...but that’s downtime
+
+The simplest way to switch writes would have been to stop the world:
+
+1. Turn off all writes.
+1. Wait for 16 to catch up
+1. Enable writes again — this time they all go to 16
+
+If we went with the ‘stop the world approach’, we’d have about the same kind of downtime as blue-green deployments: a minute or so.
+
+We were okay with a minute of downtime. But we had already spent a day setting up our manual method, could we do better?
+
+Since we were switching manually we had finer control over our connections. We realized that with just a little bit more work...we could have no downtime at all!
+
+<a name="zero-downtime-algo"></a>
+
+## Or we could write an algorithm with zero downtime!
+
+Our co-author Daniel shared an algorithm he used at his previous startup:
+
+![](/posts/pg_upgrade/no_downtime.png)
+
+First, we pause all new transactions. Then, we wait for active transactions to complete and for 16 to catch up. Finally we unpause all transactions and have them go to 16. If we did this right, we could switch major versions without any downtime at all!
+
+### The benefits of being small
+
+Sounds good in theory, but it can be hard to pull off. Unless of course you run at a modest scale.
+
+Our switching algorithm hinges on being able to control all active connections. If you have tons of machines, how could you control all active connections?
+
+Well, since our throughput was still modest, we could temporarily scale our sync servers down to just one giant machine. Clojure and java came handy here too. We had threads and the JVM is efficient, so we could take full advantage of the [m6a.16xlarge](https://instances.vantage.sh/aws/ec2/m6a.16xlarge?region=us-east-1&os=linux&cost_duration=monthly&reserved_term=Standard.noUpfront) sync server we moved too for the switch.
```

Comment:
```suggestion
Well, since our throughput was still modest, we could temporarily scale our sync servers down to just one giant machine. Clojure and java came handy here too. We had threads and the JVM is efficient, so we could take full advantage of the [m6a.16xlarge](https://instances.vantage.sh/aws/ec2/m6a.16xlarge?region=us-east-1&os=linux&cost_duration=monthly&reserved_term=Standard.noUpfront) sync server we moved to for the switch.
```